File: client/py/arya/segment/iterator.py
from dataclasses import dataclass
from enum import Enum

import freighter

from arya import telem
from arya.segment import BinarySegment

_ENDPOINT = "/segment/iterate"


class Command(int, Enum):
    OPEN = 0
    NEXT = 1
    PREV = 2
    FIRST = 3
    LAST = 4
    NEXT_SPAN = 5
    PREV_SPAN = 6
    NEXT_RANGE = 7
    SEEK_FIRST = 8
    SEEK_LAST = 9
    SEEK_LT = 10
    SEEK_GE = 11
    VALID = 12
    ERROR = 13
    CLOSE = 14
    EXHAUST = 15


class ResponseVariant(int, Enum):
    ACK = 1
    DATA = 2


@dataclass
class Request:
    command: Command
    span: telem.TimeSpan | None = None
    range: telem.TimeRange | None = None
    stamp: telem.TimeStamp | None = None
    keys: list[str] | None = None
    sync: bool = False


@dataclass
class Response:
    variant: ResponseVariant
    ack: bool
    command: Command
    error: freighter.ErrorPayload
    segments: list[BinarySegment]


def _response_factory() -> Response:
    return Response(ResponseVariant.ACK, False, Command.OPEN,
                    freighter.ErrorPayload(None, None), [])


class Core:
    transport: freighter.StreamClient
    stream: freighter.Stream[Request, Response]
    values: list[BinarySegment]

    def __init__(self, transport: freighter.StreamClient) -> None:
        self.transport = transport

    def open(self, keys: list[str], tr: telem.TimeRange):
        self.stream = self.transport.stream(_ENDPOINT, Request, _response_factory)
        self.exec(command=Command.OPEN, range=tr, keys=keys)
        self.values = []

    def exec(self, **kwargs) -> bool:
        exc = self.stream.send(Request(**kwargs))
        if exc is not None:
            raise exc
        while True:
            r, exc = self.stream.receive()
            if exc is not None:
                raise exc
            if r.variant == ResponseVariant.ACK:
                return r.ack
            self.values = r.segments

    def next(self) -> bool:
        return self.exec(command=Command.NEXT)

    def prev(self) -> bool:
        return self.exec(command=Command.PREV)

    def first(self) -> bool:
        return self.exec(command=Command.FIRST)

    def last(self) -> bool:
        return self.exec(command=Command.LAST)

    def next_span(self, span: telem.TimeSpan) -> bool:
        return self.exec(command=Command.NEXT_SPAN, span=span)

    def prev_span(self, span: telem.TimeSpan) -> bool:
        return self.exec(command=Command.PREV_SPAN, span=span)

    def next_range(self, rng: telem.TimeRange) -> bool:
        return self.exec(command=Command.NEXT_RANGE, range=rng)

    def seek_first(self) -> bool:
        return self.exec(command=Command.SEEK_FIRST)

    def seek_last(self) -> bool:
        return self.exec(command=Command.SEEK_LAST)

    def seek_lt(self, stamp: telem.TimeStamp) -> bool:
        return self.exec(command=Command.SEEK_LT, stamp=stamp)

    def seek_ge(self, stamp: telem.TimeStamp) -> bool:
        return self.exec(command=Command.SEEK_GE, stamp=stamp)

    def valid(self) -> bool:
        return self.exec(command=Command.VALID)

    def exhaust(self) -> bool:
        return self.exec(command=Command.EXHAUST)

    def close(self):
        try:
            self.exec(command=Command.CLOSE)
        except freighter.errors.EOF:
            raise ValueError("segment iterator closed unexpectedly")
        except Exception as e:
            raise e

File: core/pkg/service/framer/iterator/iterator.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package iterator

import (
	"context"

	"github.com/synnaxlabs/synnax/pkg/distribution/framer"
	"github.com/synnaxlabs/synnax/pkg/distribution/framer/frame"
	"github.com/synnaxlabs/synnax/pkg/distribution/framer/iterator"
	"github.com/synnaxlabs/x/confluence"
	"github.com/synnaxlabs/x/signal"
	"github.com/synnaxlabs/x/telem"
)

type (
	StreamIterator = framer.StreamIterator
	Request        = framer.IteratorRequest
	Response       = framer.IteratorResponse
)

const (
	AutoSpan            = iterator.AutoSpan
	CommandSeekFirst    = iterator.CommandSeekFirst
	CommandSeekLast     = iterator.CommandSeekLast
	CommandSeekLE       = iterator.CommandSeekLE
	CommandSeekGE       = iterator.CommandSeekGE
	CommandNext         = iterator.CommandNext
	CommandPrev         = iterator.CommandPrev
	CommandSetBounds    = iterator.CommandSetBounds
	ResponseVariantAck  = iterator.ResponseVariantAck
	ResponseVariantData = iterator.ResponseVariantData
	CommandError        = iterator.CommandError
	CommandValid        = iterator.CommandValid
)

type responseSegment = confluence.Segment[Response, Response]

type Iterator struct {
	requests    confluence.Inlet[Request]
	responses   confluence.Outlet[Response]
	shutdown    context.CancelFunc
	wg          signal.WaitGroup
	value       []Response
	valueFrames []frame.Frame
}

// Next reads all channel data occupying the next span of time. Returns true
// if the current IteratorServer.View is pointing to any valid segments.
func (i *Iterator) Next(span telem.TimeSpan) bool {
	i.value = i.value[:0]
	return i.exec(Request{Command: CommandNext, Span: span})
}

// Prev reads all channel data occupying the previous span of time. Returns true
// if the current IteratorServer.View is pointing to any valid segments.
func (i *Iterator) Prev(span telem.TimeSpan) bool {
	i.value = i.value[:0]
	return i.exec(Request{Command: CommandPrev, Span: span})
}

// SeekFirst seeks the Iterator the start of the Iterator range.
// Returns true if the current IteratorServer.View is pointing to any valid segments.
func (i *Iterator) SeekFirst() bool {
	i.value = i.value[:0]
	return i.exec(Request{Command: CommandSeekFirst})
}

// SeekLast seeks the Iterator the end of the Iterator range.
// Returns true if the current IteratorServer.View is pointing to any valid segments.
func (i *Iterator) SeekLast() bool {
	i.value = i.value[:0]
	return i.exec(Request{Command: CommandSeekLast})
}

// SeekLE seeks the Iterator to the first whose timestamp is less than or equal
// to the given timestamp. Returns true if the current IteratorServer.View is pointing
// to any valid segments.
func (i *Iterator) SeekLE(stamp telem.TimeStamp) bool {
	i.value = i.value[:0]
	return i.exec(Request{Command: CommandSeekLE, Stamp: stamp})
}

// SeekGE seeks the Iterator to the first whose timestamp is greater than the
// given timestamp. Returns true if the current IteratorServer.View is pointing to
// any valid segments.
func (i *Iterator) SeekGE(stamp telem.TimeStamp) bool {
	i.value = i.value[:0]
	return i.exec(Request{Command: CommandSeekGE, Stamp: stamp})
}

// Valid returns true if the Iterator is pointing at valid data and is error free.
func (i *Iterator) Valid() bool {
	return i.exec(Request{Command: CommandValid})
}

// Error returns any errors accumulated during the iterators lifetime.
func (i *Iterator) Error() error {
	_, err := i.execErr(Request{Command: CommandError})
	return err
}

// Close closes the Iterator, ensuring that all in-progress reads complete
// before closing the Source outlet. All iterators must be Closed, or the
// distribution layer will panic.
func (i *Iterator) Close() error {
	defer i.shutdown()
	i.requests.Close()
	return i.wg.Wait()
}

// SetBounds sets the lower and upper bounds of the Iterator.
func (i *Iterator) SetBounds(bounds telem.TimeRange) bool {
	return i.exec(Request{Command: CommandSetBounds, Bounds: bounds})
}

func (i *Iterator) Value() frame.Frame {
	if cap(i.valueFrames) < len(i.value) {
		i.valueFrames = make([]frame.Frame, len(i.value))
	} else {
		i.valueFrames = i.valueFrames[:len(i.value)]
	}
	for idx, v := range i.value {
		i.valueFrames[idx] = v.Frame
	}
	return frame.Merge(i.valueFrames)
}

func (i *Iterator) exec(req Request) bool {
	ok, _ := i.execErr(req)
	return ok
}

func (i *Iterator) execErr(req Request) (bool, error) {
	i.requests.Inlet() <- req
	for res := range i.responses.Outlet() {
		if res.Variant == ResponseVariantAck {
			return res.Ack, res.Error
		}
		i.value = append(i.value, res)
	}
	return false, nil
}
